aptos-move/framework/table-natives/src/lib.rs

```diff
@@ -534,14 +527,36 @@ fn get_table_handle(table: &StructRef) -> PartialVMResult<TableHandle> {
     Ok(TableHandle(handle))
 }
 
-fn serialize(layout: &MoveTypeLayout, val: &Value) -> PartialVMResult<Vec<u8>> {
-    val.simple_serialize(layout)
-        .ok_or_else(|| partial_extension_error("cannot serialize table key or value"))
+fn serialize_key(layout: &MoveTypeLayout, key: &Value) -> PartialVMResult<Vec<u8>> {
+    key.simple_serialize(layout)
+        .ok_or_else(|| partial_extension_error("cannot serialize table key"))
 }
 
-fn deserialize(layout: &MoveTypeLayout, bytes: &[u8]) -> PartialVMResult<Value> {
-    Value::simple_deserialize(bytes, layout)
-        .ok_or_else(|| partial_extension_error("cannot deserialize table key or value"))
+fn serialize_value(
+    layout_info: &LayoutInfo,
+    val: &Value,
+) -> PartialVMResult<(Bytes, Option<Arc<MoveTypeLayout>>)> {
+    let serialization_result = if layout_info.has_identifier_mappings {
```

Comment:
Orthogonal to this PR, we can remove identifier kinds which, applied on top of #11877 would allow to register your native type, and define how it should be serialized/deserialized

Reply:
So what I meant is that we can:
1. Have `MoveTypeLayout::Native(u64, layout)` and `ValueImpl::Native(u64, ...)`
2. You can register your layout as a `u64`, like `vm.register_layout(100)`.
3. When you construct a layout, you check if it is registered (if not, it's an error, i.e. kind of like an unknown tag), if so you can populate the hash set as you suggested.
4. So for any layout we know all `u64`s which represent different types of native layouts which are there.
5. Register "hooks" for serialization/deserialization like we add `NativeExtensions` right now.
6. When serializing/deserializing, we can just take the `u64` which we can use to fetch the right serializer/deserializer.

I think you mean exactly that?

Reply:
More general question, @igor-aptos if the above holds, do we need `ValueImpl::DelayedField` or `ValueImpl::Native`? Right now the use cases are only delayed so  let me know if it makes sense to rename.